src/storage-controller/src/collection_mgmt.rs

```diff
@@ -1439,46 +1435,87 @@ where
+    match status_history_desc.retention_policy {
+        StatusHistoryRetentionPolicy::LastN(n) => {
+            // BTreeMap to track the earliest events for each key.
+            let mut last_n_entries_per_key: BTreeMap<
+                K,
+                BinaryHeap<Reverse<(CheckedTimestamp<DateTime<Utc>>, Row)>>,
+            > = BTreeMap::new();
+
+            for (row, diff) in rows {
+                let datums = row.unpack();
+                let key = (status_history_desc.extract_key)(&datums);
+                let timestamp = (status_history_desc.extract_time)(&datums);
+
+                // Duplicate rows ARE possible if many status changes happen in VERY quick succession,
+                // so we go ahead and handle them.
+                assert!(
+                    diff > 0,
+                    "only know how to operate over consolidated data with diffs > 0, \
                 found diff {diff} for object {key:?} in {introspection_type:?}",
-        );
+                );
+
+                // Keep track of the timestamp of the latest row per key.
+                match latest_row_per_key.get(&key) {
+                    Some(existing) if &existing.0 > &timestamp => {}
+                    _ => {
+                        latest_row_per_key.insert(key.clone(), (timestamp, row.clone()));
+                    }
+                }
 
-        // Keep track of the timestamp of the latest row per key.
-        match latest_row_per_key.get(&key) {
-            Some(existing) if &existing.0 > &timestamp => {}
-            _ => {
-                latest_row_per_key.insert(key.clone(), (timestamp, row.clone()));
+                // Consider duplicated rows separately.
+                let entries = last_n_entries_per_key.entry(key).or_default();
+                for _ in 0..diff {
+                    // We CAN have multiple statuses (most likely Starting and Running) at the exact same
+                    // millisecond, depending on how the `health_operator` is scheduled.
+                    //
+                    // Note that these will be arbitrarily ordered, so a Starting event might
+                    // survive and a Running one won't. The next restart will remove the other,
+                    // so we don't bother being careful about it.
+                    //
+                    // TODO(guswynn): unpack these into health-status objects and use
+                    // their `Ord` impl.
+                    entries.push(Reverse((timestamp, row.clone())));
+
+                    // Retain some number of entries, using pop to mark the oldest entries for
+                    // deletion.
+                    while entries.len() > n {
+                        if let Some(Reverse((_, r))) = entries.pop() {
+                            deletions.push(r);
+                        }
+                    }
+                }
             }
         }
+        StatusHistoryRetentionPolicy::TimeWindow(time_window) => {
```

Comment:
This is the main thing that's changed. A lot of this logic was borrowed from `partially_truncate_metrics_history`